src/credentials/GroupDataProviderImpl.cpp

```diff
@@ -0,0 +1,1388 @@
+        uint16_t size = static_cast<uint16_t>(sizeof(mBuffer));
+        ReturnErrorOnFailure(storage.SyncGetKeyValue(mKey.Value(), mBuffer, size));
+
+        return CHIP_NO_ERROR;
+    }
+
+    CHIP_ERROR Load(chip::PersistentStorageDelegate & storage)
+    {
+        // Update storage key
+        UpdateKey();
+        // Load the serialized data
+        uint16_t size = static_cast<uint16_t>(sizeof(mBuffer));
+        ReturnErrorOnFailure(storage.SyncGetKeyValue(mKey.Value(), mBuffer, size));
+
+        // Decode serialized data
+        Clear();
+        TLV::TLVReader reader;
+        reader.Init(mBuffer, size);
+        ReturnErrorOnFailure(Deserialize(reader));
+
+        return CHIP_NO_ERROR;
+    }
+
+    CHIP_ERROR Delete(chip::PersistentStorageDelegate & storage)
+    {
+        // Update storage key
+        UpdateKey();
+        // Delete stored data
+        return storage.SyncDeleteKeyValue(mKey.Value());
+    }
+
+protected:
+    PersistentStorageKey mKey;
+    uint8_t mBuffer[S] = { 0 };
+    char mDebug[512];
+};
+
+struct FabricData : public PersistentData<kPersistentBufferMax>
+{
+    static const TLV::Tag kFirstEndpoint = 1;
+    static const TLV::Tag kEndpointCount = 2;
+    static const TLV::Tag kFirstKeyset   = 3;
+    static const TLV::Tag kKeysetCount   = 4;
+
+    chip::FabricIndex fabric        = 0;
+    chip::EndpointId first_endpoint = kInvalidEndpointId;
+    uint16_t endpoint_count         = 0;
+    uint16_t first_keyset           = 0;
+    uint16_t keyset_count           = 0;
+
+    FabricData(chip::FabricIndex fabric_index) : fabric(fabric_index) {}
+
+    void UpdateKey() override { this->mKey.FabricGroups(this->fabric); }
+
+    void Clear() override
+    {
+        first_endpoint = kInvalidEndpointId;
+        first_keyset   = 0;
+        keyset_count   = 0;
+    }
+
+    CHIP_ERROR Serialize(TLV::TLVWriter & writer) const override
+    {
+        ReturnErrorOnFailure(writer.Put(kFirstEndpoint, static_cast<uint16_t>(first_endpoint)));
+        ReturnErrorOnFailure(writer.Put(kEndpointCount, static_cast<uint16_t>(endpoint_count)));
+        ReturnErrorOnFailure(writer.Put(kFirstKeyset, static_cast<uint16_t>(first_keyset)));
+        ReturnErrorOnFailure(writer.Put(kKeysetCount, static_cast<uint16_t>(keyset_count)));
+
+        return CHIP_NO_ERROR;
+    }
+    CHIP_ERROR Deserialize(TLV::TLVReader & reader) override
+    {
+        // first_endpoint
+        ReturnErrorOnFailure(reader.Next(kFirstEndpoint));
+        ReturnErrorOnFailure(reader.Get(first_endpoint));
+        // endpoint_count
+        ReturnErrorOnFailure(reader.Next(kEndpointCount));
+        ReturnErrorOnFailure(reader.Get(endpoint_count));
+        // first_keyset
+        ReturnErrorOnFailure(reader.Next(kFirstKeyset));
+        ReturnErrorOnFailure(reader.Get(first_keyset));
+        // keyset_count
+        ReturnErrorOnFailure(reader.Next(kKeysetCount));
+        ReturnErrorOnFailure(reader.Get(keyset_count));
```

Comment:
This assumes the data is in order. A future version with different data order could fail. This can be fixed later with some work on utilities to simplify the common/easy cases.

Reply:
I am aware. I considered looping through all the keys, and set the corresponding field based on the tags, but:
* This problem can be better fixed with the addition of a get-by-tag function.
* I think the get-by-tag utility should be part of the TLV interface, so adding it here seems wasteful.
* I felt this file is rather long, and complex enough, so I decided to keep it simple for now.